pkg/kubectl/cmd/drain.go

```diff
@@ -266,40 +267,36 @@ func (o *DrainOptions) deleteOrEvictPodsSimple() error {
 	return err
 }
 
-func (o *DrainOptions) getController(sr *api.SerializedReference) (interface{}, error) {
-	switch sr.Reference.Kind {
+func (o *DrainOptions) getController(pod *api.Pod, controllerRef *metav1.OwnerReference) (interface{}, error) {
+	switch controllerRef.Kind {
 	case "ReplicationController":
-		return o.client.Core().ReplicationControllers(sr.Reference.Namespace).Get(sr.Reference.Name, metav1.GetOptions{})
+		return o.client.Core().ReplicationControllers(pod.Namespace).Get(controllerRef.Name, metav1.GetOptions{})
 	case "DaemonSet":
-		return o.client.Extensions().DaemonSets(sr.Reference.Namespace).Get(sr.Reference.Name, metav1.GetOptions{})
+		return o.client.Extensions().DaemonSets(pod.Namespace).Get(controllerRef.Name, metav1.GetOptions{})
 	case "Job":
-		return o.client.Batch().Jobs(sr.Reference.Namespace).Get(sr.Reference.Name, metav1.GetOptions{})
+		return o.client.Batch().Jobs(pod.Namespace).Get(controllerRef.Name, metav1.GetOptions{})
 	case "ReplicaSet":
-		return o.client.Extensions().ReplicaSets(sr.Reference.Namespace).Get(sr.Reference.Name, metav1.GetOptions{})
+		return o.client.Extensions().ReplicaSets(pod.Namespace).Get(controllerRef.Name, metav1.GetOptions{})
 	case "StatefulSet":
-		return o.client.Apps().StatefulSets(sr.Reference.Namespace).Get(sr.Reference.Name, metav1.GetOptions{})
+		return o.client.Apps().StatefulSets(pod.Namespace).Get(controllerRef.Name, metav1.GetOptions{})
 	}
-	return nil, fmt.Errorf("Unknown controller kind %q", sr.Reference.Kind)
+	return nil, fmt.Errorf("Unknown controller kind %q", controllerRef.Kind)
 }
 
-func (o *DrainOptions) getPodCreator(pod api.Pod) (*api.SerializedReference, error) {
-	creatorRef, found := pod.ObjectMeta.Annotations[api.CreatedByAnnotation]
-	if !found {
+func (o *DrainOptions) getPodCreator(pod api.Pod) (*metav1.OwnerReference, error) {
+	controllerRef := controller.GetControllerOf(&pod)
+	if controllerRef == nil {
 		return nil, nil
 	}
-	// Now verify that the specified creator actually exists.
-	sr := &api.SerializedReference{}
-	if err := runtime.DecodeInto(o.Factory.Decoder(true), []byte(creatorRef), sr); err != nil {
-		return nil, err
-	}
+
 	// We assume the only reason for an error is because the controller is
 	// gone/missing, not for any other cause.  TODO(mml): something more
 	// sophisticated than this
-	_, err := o.getController(sr)
+	_, err := o.getController(&pod, controllerRef)
```

Comment:
Actually, do we still need this? This originally checks if the controller specified in created-by annotation is gone or missing. Created-by annotation is never updated, but controller ref is different. If a pod has a controller ref set, the controller must exist. @enisoc